onnxruntime/core/framework/allocation_planner.cc

```diff
@@ -585,6 +588,53 @@ class PlannerImpl {
     return Status::OK();
   }
 
+  // Whether a given NodeArg has fence or not.
+  // If the buffer is reused, need to check whether original OrtValue has fence or not.
+  bool HasFence(const onnxruntime::NodeArg* arg) {
+    bool has_fence = false;
+    if (arg && arg->Exists()) {
+      OrtValueIndex index = Index(arg->Name());
+      AllocPlanPerValue& value_plan = AllocPlan(index);
+      if (value_plan.alloc_kind == AllocKind::kReuse)
+      {
+        // Buffer reused, check original buffer to see if fence is shared.
+        has_fence = AllocPlan(value_plan.reused_buffer).create_fence_if_async;
+      }
+      else
+      {
+        has_fence = value_plan.create_fence_if_async;
+      }
+    }
+
+    return has_fence;
+  }
+
+  // Compute fence check. Set has_fence flag if either one of inputs, implicit inputs or outputs of a given node has fence.
+  Status ComputeFenceCheck() {
+
+    for (SequentialExecutionPlan::NodeExecutionPlan& step : plan_.execution_plan) {
+      auto pnode = graph_viewer_.GetNode(step.node_index);
+      if (pnode == nullptr) return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "Can not find the node ", step.node_index);
+
+      bool has_fence = false;
+      for (auto node_input : pnode->InputDefs()) {
+        has_fence |= HasFence(node_input);
```

Comment:
bit wise or? maybe cleaner to use logic or.

Reply:
bit wise or is recommended for boolen instead of logic or.

Reply:
> bit wise or? maybe cleaner to use logic or.

https://stackoverflow.com/questions/2488406/why-doesnt-c-have-or-for-booleans

Conclusion
Therefore YES we can replace && by & for bool values ;-) 
So better use &= instead of &&=. 
We can consider &&= as useless for booleans.

Reply:
Not sure about the recommendation. Logic or can shortcut to save some perf, while bitwise or cannot.

---
In reply to: [312212948](https://github.com/microsoft/onnxruntime/pull/1593#discussion_r312212948) [](ancestors = 312212948)

Reply:
> Not sure about the recommendation. Logic or can shortcut to save some perf, while bitwise or cannot.
> 
> In reply to: 312212948 [](ancestors = 312212948)

Agree with you logic or might be able to save some cost on HasFence(). Will change it.